body.go

```diff
@@ -0,0 +1,64 @@
+package main
+
+import (
+	"errors"
+	"fmt"
+	"io"
+	"os"
+)
+
+type body struct {
+	File   *os.File
+	Reader io.Reader
+	Offset int64
+	Size   int64
+}
+
+// Create a body.  A body is an io.Reader instance which reads from the file at
+// filename, starting at the `offset`th byte and reading up to `size` bytes in
+// total.
+func newBody(filename string, offset, size int64) (*body, error) {
+	if size == 0 {
+		return nil, errors.New("Cannot specify a size of 0")
```

Comment:
Unfortunately errors in go tend to get propagated "up the stack" so sometimes client code might not realise where the error originated from. In this case it might be better to state something like `fmt.Errorf("taskcluster-lib-artifact-go: artifact.NewBody called with size 0 for filename %v and offset %v - size 0 is not allowed.", filename, offset)`.

Reply:
:'( I'm sure not a fan of error handling in Go, but that's a good suggestion.

Reply:
Yeah I am also not the greatest fan of error handling in go.

Another (slightly more involved) option is to create a custom type for the error, e.g.:

```go
type NilBodySizeError {
    Filename string
    Offset int64
}

func (n *NilBodySizeError) Error() string {
    return fmt.Sprintf("NilBodySizeError for filename %v and offset %v", n.Filename, n.Offset)
}

// NewBody creates a new artifact body ....
// If size is 0, an error will be returned of type *NilBodySizeError.
func NewBody(filename string, offset, size int64) (body *Body, err error) {
    if size == 0 {
        return nil, NilBodySizeError {
            Filename: filename,
            Offset: offset,
        }
    }
    ....
}
```

The advantage of this is that calling code sees the error is of type `*artifact.NilBodySizeError` and so knows which package this error comes from, and can also directly see the doc comments for that error type. Also calling code can query the members of the error (`Filename` and `Offset`) and do custom handling based on the values, if it really wanted to.